### PR TITLE
[POC] Add possibility to switch configuration 

### PR DESCRIPTION
**Pull request by ismail1432**
Following https://github.com/jolicode/elastically/issues/22 I poc a simple way to switch configuration, by default we always take configuration from yaml file but we can also define a custom configuration class. 

```yaml
    JoliCode\Elastically\Client:
        arguments:
            $config:
                configuration_class: '@App\Search\DudeConfiguration' 
```

```diff
+  configuration_class: '@App\Search\DudeConfiguration' 
```

`App\Search\DudeConfiguration` can be define as a service so you can inject any others service needed. Suppose that I need to fecth configuration from a database

```php
class DudeConfiguration extends AbstractConfiguration
{
    public function __construct(array $options = [], EntityManagerInterface $manager)
    {
        $config = [
            'elastically_mappings_directory' => $manager->getRepository(Config::class)->findOneBy(['configName'=> 'path']),
            'elastically_bulk_size' => $manager->getRepository(Config::class)->findOneBy(['configName'=>'bulkSize'])
            ];
   
        parent::__construct($config);
    }
}
```
At the moment we not handle the mapping definition but 
I think we can add a option `elastically_mapping` that return the whole mapping and in `IndexBuilder::createIndex` we get it the mapping from the option resolver. 
Something like :

```diff
- $mappingFilePath = $this->configuration->getOption('elastically_mappings_directory').DIRECTORY_SEPARATOR.$indexName.'_mapping.yaml';
- $mapping = Yaml::parseFile($mappingFilePath);
+ $mapping = $this->configuration->getOption('elastically_mappings')
```

### CODE REVIEW

File: src/Client.php [hunk 1]
```diff
@@ -25,10 +25,13 @@ class Client extends ElasticaClient
     private $indexer;
     private $indexBuilder;
 
+    /** @var AbstractConfiguration $configuration */
+    private $configuration;
```
Comment (damienalexandre) [inline src/Client.php:29]: I thinks what you wanted to achieve here is more a "MappingReader" than a "Configuration"; configuration is still an array given to the Client and this new AbstractConfiguration class looks ambiguous.

File: src/Client.php [hunk 2]
```diff
@@ -25,10 +25,13 @@ class Client extends ElasticaClient
...
+
     public function getIndexBuilder(): IndexBuilder
     {
         if (!$this->indexBuilder) {
-            $this->indexBuilder = new IndexBuilder($this, $this->getConfig(self::CONFIG_MAPPINGS_DIRECTORY));
+            $this->indexBuilder = new IndexBuilder($this, $this->getConfiguration());
```
Comment (damienalexandre) [inline src/Client.php:34]: Maybe having "mapping reader" is better for understanding the code here?

```suggestion
            $this->indexBuilder = new IndexBuilder($this, $this->getMappingReader());
```

File: src/Client.php (hunk 1 above)
Comment (ismail1432) [reply src/Client.php]: I do agree I updated in this way

File: src/Client.php (hunk 2 above)
Comment (ismail1432) [reply src/Client.php]: fixed

**Comment by ismail1432** [comment]
> Some comments about naming things :)
> 
> Also could you fix the tests?
> 
> Thanks a lot

Thanks for the feedback, I updated the code with your comments,
and I also replace the logic to resolve the configuration to get them from mapping reader **that still be from the yaml by default** 
Exemple :

```diff
- $indexToClass = $this->getConfig(Client::CONFIG_INDEX_CLASS_MAPPING);
+ $indexToClass = $this->getMappingReader()->getOption(self::CONFIG_INDEX_CLASS_MAPPING);
```

I'm still open if there is something wrong or anything to improve and OFC I'll fix test :+1:

**Comment by ismail1432** [comment]
Hi @damienalexandre,

Sorry for the delay, I'll close this one as I went too far

**Pull request closed without merging**